### PR TITLE
Add Apple translation popovers for notes for iOS 17.4+ and macOS 14.4+

### PR DESCRIPTION
**Pull request by tyiu**
Changelog-Added: Add Apple translation popovers for notes for iOS 17.4+ and macOS 14.4+

https://github.com/damus-io/damus/assets/963907/4fdddb61-9cd0-4514-80fe-79298d9c92a6

Simulator Screenshot - iPhone 15 Pro - 2024-07-16 at 13 07 06 Medium

These changes are gated around iOS and macOS versions. If an older OS is being run, Apple translations won't be offered.

### CODE REVIEW

**Comment by alltheseas** [comment]
@tyiu is it possible to set up auto translate using iOS offline translations? That is, remove the need to tap "translate  note"

**Comment by tyiu** [comment]
> @tyiu is it possible to set up auto translate using iOS offline translations? That is, remove the need to tap "translate  note"

Technically yes, but only in iOS 18.0+

**Comment by jb55** [comment]
very cool, will test this soon

**Comment by tyiu** [comment]
There's a bug when `damus_state.settings.auto_translate` is set to true but `Offline Translations` is selected. It won't show the Translate button in that case, which is incorrect. I'll work on a fix, and also look into the auto translations with iOS 18.0.

**Comment by tyiu** [comment]
I mischaracterized the this PR. It's not exclusively offline note translations from Apple. The translation requests can still go to Apple servers, but can be offline as well via downloaded languages if there's network connectivity issues. I've updated the description, commit, and PR title accordingly

>There's a bug when damus_state.settings.auto_translate is set to true but Offline Translations is selected. It won't show the Translate button in that case, which is incorrect. I'll work on a fix, and also look into the auto translations with iOS 18.0
I fixed the bug that I reported in my previous comment.

File: damus/Models/TranslationService.swift [hunk 1]
```diff
@@ -38,7 +38,13 @@ enum TranslationService: String, CaseIterable, Identifiable, StringCodable {
     var model: Model {
         switch self {
         case .none:
-            return .init(tag: self.rawValue, displayName: NSLocalizedString("none_translation_service", value: "None", comment: "Dropdown option for selecting no translation service."))
+            let displayName: String
+            if TranslationService.isAppleTranslationPopoverSupported {
```
Comment (tyiu) [inline damus/Models/TranslationService.swift:42]: I'd like to solicit feedback here. Right now, if you have iOS 17.4+, there's no way to turn off translations. It's just using Apple by default, but requires an explicit tap on the Translate Note button.

Alternatively, I could make `None` and `Apple` separate options, but it feels like we have an opportunity to be more opinionated here to just say everyone needs translations.

Comment (danieldaquino) [reply damus/Models/TranslationService.swift:42]: I think we need to consider the privacy aspect, given that some of those translate requests might be sent to Apple servers.

In my opinion, we should be opinionated and place the button "Translate note" by default, and then add a confirmation dialog explaining the user the text _might_ be sent to Apple servers, with an option to dismiss that dialog forever if the user chooses to do so.

For the iOS 18 feature where translations can be made automatically, I believe that should be opt-in.

Comment (tyiu) [reply damus/Models/TranslationService.swift:42]: > I think we need to consider the privacy aspect, given that some of those translate requests might be sent to Apple servers.
> 
> In my opinion, we should be opinionated and place the button "Translate note" by default, and then add a confirmation dialog explaining the user the text _might_ be sent to Apple servers, with an option to dismiss that dialog forever if the user chooses to do so.

Simulator Screenshot - iPhone 16 Pro - 2024-09-22 at 00 44 41

It turns out that Apple shows exactly that confirmation dialog on the first attempt, so I think we're OK there. Agreed about the automatic iOS 18 translations feature, though.

**Comment by tyiu** [comment]
I rebased this branch on the latest master and force pushed. No other changes were made.

File: damus/Models/TranslationService.swift (hunk 1 above)
Comment (danieldaquino) [reply damus/Models/TranslationService.swift:42]: Thank you @tyiu, I will take a second look at this PR soon with this in mind!

**Review by danieldaquino — APPROVED** [review]
Thank you @tyiu, reviewed and tested it!

- Default settings and the overall flow makes a lot of sense
- Works well _(Tested on iPhone 13 mini on iOS 18.0 switching between Damus Purple translation and Apple)_
- I really like how clear and readable the code changes are! Good job!

Merging!

**Pull request merged**